New ASM class for (linear) Lagrange meshes from file

kmokstad opened the pull request:
The added class ASMs2DMatlab is a subclass of ASM2DLag and reads the mesh file dumped by the SIMoutput:dumpMatlabGrid method, which corresponds to the mesh-files from the S_Miehe matlab solver. All this for the convenience of comparing the two solvers on the same grid.

VikingScientist commented:
Rename to avoid calling it 'matlab'

akva2 commented on src/ASM/ASMs2DMatlab.C:
```diff
@@ -0,0 +1,211 @@
+        std::cout << id <<": "<< X << std::endl;
+#endif
+	this->setCoord(id,X);
+        if (cline.find_last_of(';') == cline.length()-1) break;
```
cline.back() == ';'

kmokstad replied:
fixed

akva2 commented on src/ASM/ASMs2DMatlab.C:
```diff
@@ -0,0 +1,211 @@
+
+	int node;
+	IntVec& nodes = nodeSets[setname];
+        std::istringstream selem(cline);
```
indent

kmokstad replied:
was some left-over tabs inserted by emacs, cleaned now

akva2 commented on src/ASM/ASMs2DMatlab.C:
```diff
@@ -0,0 +1,211 @@
+{
+  p1 = p2 = 2; // So far only linear elements supported
+
+  nx = sqrt(nnod);
```
nx = ny  = sqrt(nnod) ?

kmokstad replied:
Doesn't really matter, as they are not used anyway when the mesh is unstructured. Removed.

akva2 commented on src/ASM/ASMs2DMatlab.C:
```diff
@@ -0,0 +1,211 @@
+{
+  std::map<std::string,IntVec>::const_iterator it = nodeSets.begin();
+  for (int i = 1; it != nodeSets.end(); i++, ++it)
+    if (it->first == setName) return i;
```
strictly speaking i may be invalidated by additional inserts (not going to happen so just a knee-jerk).

kmokstad replied:
True, but the assumption is that the nodeSets is defined only by the read method, and after that left untouched. Therefore the access method only returns a const reference to it.

akva2 commented on src/ASM/ASMs2DMatlab.C:
```diff
@@ -0,0 +1,211 @@
+  threadGroups[0].resize(1);
+  threadGroups[1].resize(0);
+  threadGroups[0][0].resize(nel);
+  std::iota(threadGroups[0][0].begin(),threadGroups[0][0].end(),0);
```
maybe we should consider sticking this in base ?

kmokstad replied:
yes

akva2 commented on src/ASM/ASMbase.C:
```diff
@@ -509,6 +509,20 @@ void ASMbase::constrainPatch (int dof, int code)
 }
 
 
+void ASMbase::constrainNodes (const IntVec& nodes, int dof, int code)
```
nice. we can probably use this to simplify some other methods at some point.

kmokstad commented:
Renamed the new class to ASMu2DLag instead, as suggested, and added the matlab-syntax reading as a separate method. Then, it can be extended by other formats when needed.

kmokstad commented:
jenkins build this with downstreams please

akva2 commented on src/ASM/ASMu2DLag.C:
```diff
@@ -195,24 +196,38 @@ bool ASMu2DLag::generateFEMTopology ()
+    if (nset.first == setName)
+      return nset.second;
+
+  nodeSets.push_back(std::make_pair(setName,IntVec()));;
```
maybe std::unordered_map is more convenient ?
